design/sticky-session-design.md

```diff
@@ -0,0 +1,170 @@
+# Sticky session support
+
+**Status**: _draft_
+
+This document proposes adding support for sticky session in contour.
+
+## Goals
+
+- Enabling support for sticky session in Contour using
+    - Cookie
+    - Source IP
+    - Header
+- Adding support in IngressRoute CRD. 
+
+## Non-goals
+
+- Making options in IngressRoute very specific to Envoy is not a goal. Options in IngressRoute will be generic.
+
+## Background
+
+Sticky-sessions allow requests to be routed to the same host based on certain parameters of the request.
+The parameters can be a cookie, the source IP, a custom header, etc.
+
+Envoy supports sticky sessions by using a hash-based load balancing strategy (RingHash, Maglev). 
+Specifying a *hash policy* for such load-balancers will enable sticky sessions. However, we are not aiming at making
+this very specific to Envoy. IngressRoute options should be generic enough without exposing
+internals of Envoy itself.
+
+
+## High-Level Design
+
+Sticky sessions are enabled per route. The following change are required in IngressRoute:
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata: 
+  name: basic
+  namespace: default
+spec: 
+  virtualhost:
+    fqdn: foo-basic.bar.com
+  routes: 
+    - match: /
+      sticky-session:
+        source: cookie/header/connection # There should be exactly one of these set.
```

Comment:
This is good but I think we can tighten this up quite a bit using different data structures to represent the different options; that also might mean we don't have to implement each of these options initailly.

My suggestion is 

```
routes:
  - match: /
     sticky-session:
       cookie:
          name: // name of cookie
          path: // cookie path
```
Then that opens the door for coming back and adding a the other types under their own key

Reply:
Sure, this seems right.